File: src/game.ml
open Card
open Trick
open Hand

type task_status = Done | Failed | Unknown

type player = {
  hand : hand;
  tricks : trick Dynarray.t;
  tasks : task Dynarray.t;
}

and state = {
  players : player array;
  mutable current_trick : trick;
  mutable first_player : int;
}

and task = state -> task_status

let is_captain player =
  let is_trump_4 c = c == Trump 4 in
  Dynarray.exists is_trump_4 player.hand
  || Dynarray.exists (fun t -> Dynarray.exists is_trump_4 t.cards) player.tricks

let retrieve_captain state =
  let ic = ref 0 in
  Array.iteri (fun i p -> if is_captain p then ic := i) state.players;
  !ic

let n_players state = Array.length state.players

let current_player_index state =
  (state.first_player + Dynarray.length state.current_trick.cards)
  mod n_players state

let current_player state = state.players.(current_player_index state)

let game_failed state =
  let task_failed t = t state == Failed in
  let player_failed p = Dynarray.exists task_failed p.tasks in
  Array.exists player_failed state.players

let game_won state =
  let task_done t = t state == Done in
  let player_done p = Dynarray.for_all task_done p.tasks in
  Array.for_all player_done state.players

let game_over state =
  let cards_left =
    Array.fold_left (fun x p -> x + Dynarray.length p.hand) 0 state.players
  in
  nb_cards mod n_players state == cards_left

File: rust/the-crew-solver/src/game.rs
use crate::{player::Player, task::TaskStatus, trick::Trick};

#[derive(Debug, Clone)]
pub struct State {
    players: Vec<Player>,
    first_player: usize,
    current_trick: Trick,
}

#[derive(Debug)]
pub enum GameError {
    MissingCaptain,
    InvalidTrickSize(usize, usize),
}

impl State {
    pub fn retrieve_captain(players: &Vec<Player>) -> Result<usize, GameError> {
        for (i, p) in players.iter().enumerate() {
            if p.is_captain() {
                return Ok(i);
            }
        }
        Err(GameError::MissingCaptain)
    }

    pub fn new(players: Vec<Player>) -> Self {
        let captain = State::retrieve_captain(&players).unwrap();
        State {
            players,
            first_player: captain,
            current_trick: (0, vec![]).into(),
        }
    }

    pub fn n_players(&self) -> usize {
        self.players.len()
    }

    pub fn get_player(&self, i: usize) -> &Player {
        &self.players[i]
    }

    pub fn current_player_idx(&self) -> usize {
        (self.first_player + self.current_trick.len()) % self.n_players()
    }

    pub fn current_player(&self) -> &Player {
        &self.players[self.current_player_idx()]
    }

    pub fn game_status(&self) -> TaskStatus {
        let mut done = true;
        for (i, p) in self.players.iter().enumerate() {
            match p.task_status(i, self) {
                TaskStatus::Failed => return TaskStatus::Failed,
                TaskStatus::Unknown => done = false,
                TaskStatus::Done => (),
            }
        }
        if done {
            TaskStatus::Done
        } else {
            TaskStatus::Unknown
        }
    }
}
